c10/cuda/impl/CUDAGuardImpl.h

```diff
@@ -59,7 +62,9 @@ struct CUDAGuardImpl final : public c10::impl::DeviceGuardImplInterface {
   void uncheckedSetDevice(Device d) const noexcept override {
     auto current_device = uncheckedGetDevice();
     if (!current_device.has_value() || current_device.value() != d) {
-      C10_CUDA_CHECK_WARN(cudaSetDevice(d.index()));
+      if (driver_api->hasPrimaryContext(d.index())) {
```

Comment:
suppose you have the following snippet:
```
x=torch.randn(1, device="cuda:1")
y=torch.randn(1, device="cuda")
```
The first line will create DeviceGuard for device 1, but when that device guard is destroyed, current device won't be reset to 0, and will still be 1. Then second line will create `y` on current device, which is 1, instead of 0, as expected?

Reply:
Yes, this is exactly what will happen. In addition to that, the following code snippet:
```Python
torch.cuda.set_device(0)
x=torch.randn(1, device="cuda:1")
y=torch.randn(1, device="cuda")
```
will allocate `x` on device 1 and `y` on device 0.

Reply:
Wait, will `torch.cuda.set_device(0)` create context on device 0? That's not what happens today. In addition, behavior I described in my comment is also not what happens today.

Reply:
Sorry, you are right. That comment was about CUDA 12. 
On CUDA 11, even if we set device 0 the beginning
```Python
torch.cuda.set_device(0)
x=torch.randn(1, device="cuda:1") # allocated on cuda:1
y=torch.randn(1, device="cuda") # allocated on cuda:1
```
It will allocate both tensors on 1. 
To achieve the old behavior the context should be initialized:
```Python
torch.cuda.set_device(0)
foo = torch.randn(1, device="cuda")
x=torch.randn(1, device="cuda:1") # allocated on cuda:1
y=torch.randn(1, device="cuda") # allocated on cuda:0
```

Reply:
I'm sorry, this won't do. It is severely bc breaking.